cores/esp32/IPAddress.cpp

```diff
@@ -364,5 +244,168 @@ bool IPAddress::fromString6(const char *address) {
+    return *this;
+}
+
+bool IPAddress::operator==(const IPAddress& addr) const {
+    return (addr._type == _type)
+        && (memcmp(addr._address.bytes, _address.bytes, sizeof(_address.bytes)) == 0);
+}
+
+bool IPAddress::operator==(const uint8_t* addr) const
+{
+    // IPv4 only comparison to byte pointer
+    // Can't support IPv6 as we know our type, but not the length of the pointer
+    return _type == IPv4 && memcmp(addr, &_address.bytes[IPADDRESS_V4_BYTES_INDEX], sizeof(uint32_t)) == 0;
+}
+
+uint8_t IPAddress::operator[](int index) const {
+    if (_type == IPv4) {
+        return _address.bytes[IPADDRESS_V4_BYTES_INDEX + index];
+    }
+    return _address.bytes[index];
+}
+
+uint8_t& IPAddress::operator[](int index) {
+    if (_type == IPv4) {
+        return _address.bytes[IPADDRESS_V4_BYTES_INDEX + index];
+    }
+    return _address.bytes[index];
+}
+
+size_t IPAddress::printTo(Print& p) const
+{
+    size_t n = 0;
+
+    if (_type == IPv6) {
+        // IPv6 IETF canonical format: compress left-most longest run of two or more zero fields, lower case
+        int8_t longest_start = -1;
+        int8_t longest_length = 1;
+        int8_t current_start = -1;
+        int8_t current_length = 0;
+        for (int8_t f = 0; f < 8; f++) {
+            if (_address.bytes[f * 2] == 0 && _address.bytes[f * 2 + 1] == 0) {
+                if (current_start == -1) {
+                    current_start = f;
+                    current_length = 1;
+                } else {
+                    current_length++;
+                }
+                if (current_length > longest_length) {
+                    longest_start = current_start;
+                    longest_length = current_length;
+                }
+            } else {
+                current_start = -1;
+            }
+        }
+        for (int f = 0; f < 8; f++) {
+            if (f < longest_start || f >= longest_start + longest_length) {
+                uint8_t c1 = _address.bytes[f * 2] >> 4;
+                uint8_t c2 = _address.bytes[f * 2] & 0xf;
+                uint8_t c3 = _address.bytes[f * 2 + 1] >> 4;
+                uint8_t c4 = _address.bytes[f * 2 + 1] & 0xf;
+                if (c1 > 0) {
+                    n += p.print((char)(c1 < 10 ? '0' + c1 : 'a' + c1 - 10));
+                }
+                if (c1 > 0 || c2 > 0) {
+                    n += p.print((char)(c2 < 10 ? '0' + c2 : 'a' + c2 - 10));
+                }
+                if (c1 > 0 || c2 > 0 || c3 > 0) {
+                    n += p.print((char)(c3 < 10 ? '0' + c3 : 'a' + c3 - 10));
+                }
+                n += p.print((char)(c4 < 10 ? '0' + c4 : 'a' + c4 - 10));
+                if (f < 7) {
+                    n += p.print(':');
+                }
+            } else if (f == longest_start) {
+                if (longest_start == 0) {
+                    n += p.print(':');
+                }
+                n += p.print(':');
+            }
+        }
```

Comment:
It is a bit elaborate code.
Maybe something like this will result in smaller code?

```c++
        for (int f = 0; f < 8; f++) {
            if (f < longest_start || f >= longest_start + longest_length) {
                uint16_t w = (static_cast<uint16_t>(_address.bytes[f * 2]) << 8) |
                             _address.bytes[f * 2 + 1];

                bool add = false;
                for (int i = 0; i < 4; ++i) {
                    const uint8_t c = (w & 0xf000);
                    if (c > 0) add = true;
                    if (add) {
                        n += p.print((char)(c < 10 ? '0' + c : 'a' + c - 10));
                    }
                    w <<= 4;
                }
                if (!add)
                    n += p.print('0');

                if (f < 7) {
                    n += p.print(':');
                }
            } else if (f == longest_start) {
                if (longest_start == 0) {
                    n += p.print(':');
                }
                n += p.print(':');
            }
        }
```

(not yet tested, just a quick dump of an idea I had)

Reply:
Sure, slightly shorter code (and an optimiser might unroll the loop anyway), but linear code is usually not elaborate (long, but not complex).

Still, I think it is reasonable to put into a loop.

This is a place where unit tests would be really good, to check nothing breaks. I do suggest adding them.

Also, I would suggest contributing this change back to ArduinoCore, where the algorithm was first added. Note that ArduinoCore does have a lot of unit tests, so making the change there and running the unit tests would be a good thing to validate (even if not accepted by them yet).

Reply:
Yuop. agree here. @TD-er everything in IPAddress, except `zone`, `to_ip_addr_t` and `from_ip_addr_t` is carbon copy of Arduino's official core API. And as @sgryphon said, it should be first fixed/shortened there.